src/main/java/com/github/manolo8/darkbot/core/MemoryAPI.java

```diff
@@ -0,0 +1,428 @@
+package com.github.manolo8.darkbot.core;
+
+import eu.darkbot.api.API;
+
+/**
+ * Provides access to read/write native memory
+ * <p>
+ * Those calls should be generally safe,
+ * every access violation error is handled by native code
+*/
+@SuppressWarnings({"PMD.TooManyMethods", "PMD.ExcessivePublicCount", "unused"})
+public interface MemoryAPI extends API.Singleton {
+    long NULL = 0;
+    long ATOM_KIND = 0b111L;
+    long ATOM_MASK = ~ATOM_KIND;
+
+    /**
+     * String used as fallback for {@link #readString(long)} when returned value is null.
+     */
+    String FALLBACK_STRING = "ERROR";
+
+    /**
+     * Reads signed integer value from memory.
+     *
+     * @param address to read
+     * @return signed integer value
+     */
+    int readInt(long address);
+
+    default int readInt(long address, int o1) {
+        return readInt(address + o1);
+    }
+
+    default int readInt(long address, int o1, int o2) {
+        return readInt(readLong(address, o1) + o2);
+    }
+
+    default int readInt(long address, int o1, int o2, int o3) {
+        return readInt(readLong(address, o1, o2) + o3);
+    }
+
+    default int readInt(long address, int o1, int o2, int o3, int o4) {
+        return readInt(readLong(address, o1, o2, o3) + o4);
+    }
+
+    default int readInt(long address, int... offsets) {
+        int i = 0;
+        for (; i < offsets.length - 1; i++) {
+            address = readLong(address + offsets[i]);
+        }
+
+        return readInt(address + offsets[i]);
+    }
+
+    /**
+     * Reads signed long value from memory.
+     *
+     * @param address to read
+     * @return signed long value
+     */
+    long readLong(long address);
+
+    default long readLong(long address, int o1) {
+        return readLong(address + o1);
+    }
+
+    default long readLong(long address, int o1, int o2) {
+        return readLong(readLong(address, o1) + o2);
+    }
+
+    default long readLong(long address, int o1, int o2, int o3) {
+        return readLong(readLong(address, o1, o2) + o3);
+    }
+
+    default long readLong(long address, int o1, int o2, int o3, int o4) {
+        return readLong(readLong(address, o1, o2, o3) + o4);
```

Comment:
shouldn't this be like:
```suggestion
        return readLong(readAtom(address, o1, o2, o3) + o4);
```
?